specs/_features/execution_proofs/beacon-chain.md

```diff
@@ -0,0 +1,214 @@
+
+## Configuration
+
+*Note*: The configuration values are not definitive.
+
+| Name                             | Value                           |
+| -------------------------------- | ------------------------------- |
+| `MIN_REQUIRED_EXECUTION_PROOFS`  | `uint64(1)`                     |
+
+## Containers
+
+### New containers
+
+#### `ExecutionProof`
+
+```python
+class ExecutionProof(Container):
+    beacon_root: Root
+    zk_proof: ZKProof
+    validator_index: ValidatorIndex
+```
+
+#### `SignedExecutionProof`
+
+```python
+class SignedExecutionProof(Container):
+    message: ExecutionProof
+    signature: BLSSignature
+```
+
+### Extended Containers
+
+*Note*: `BeaconState` and `BeaconBlockBody` remain the same as Fulu. No modifications are required for execution proofs since they are handled externally.
+
+## Helper functions
+
+### Execution proof functions
+
+#### `get_el_program`
+
+```python
+def get_el_program(proof_id: ProofID) -> EL_PROGRAM:
+    """
+    Get the EL program for a given proof system ID.
+    """
+    if proof_id == ProofID(0):
+        return RETH_EL_PROGRAM
+    elif proof_id == ProofID(1):
+        return GETH_EL_PROGRAM
+    elif proof_id == ProofID(2):
+        return NETHERMIND_EL_PROGRAM
+    elif proof_id == ProofID(3):
+        return BESU_EL_PROGRAM
+
+    raise Exception(f"Unsupported proof_id: {proof_id}") 
+```
+
+#### `verify_execution_proof`
+
+```python
+def verify_execution_proof(proof: ExecutionProof, parent_hash: Hash32, block_hash: Hash32, state: BeaconState, el_program: EL_PROGRAM) -> bool:
+    """
+    Verify an execution proof against a payload header using zkEVM verification.
+    """
+
+    # Note: proof.beacon_root verification would be done at a higher level
+
+    # Verify the validator signature
+    validator = state.validators[proof.validator_index]
+    signing_root = compute_signing_root(proof, get_domain(state, DOMAIN_EXECUTION_PROOF))
+    if not bls.Verify(validator.pubkey, signing_root, proof.signature):
+        return False
+
+    return verify_zkevm_proof(proof.zk_proof, parent_hash, block_hash, el_program)
+```
+
+#### `verify_execution_proofs`
+
+```python
+def verify_execution_proofs(parent_hash: Hash32, block_hash: Hash32, state: BeaconState) -> bool:
+    """
+    Verify that execution proofs are available and valid for an execution payload.
+    """
+    # `retrieve_execution_proofs` is implementation and context dependent
+    # It returns all the execution proofs for the given payload block hash
+    execution_proofs = retrieve_execution_proofs(block_hash)
+
+    # Verify we have sufficient proofs
+    if len(execution_proofs) < MIN_REQUIRED_EXECUTION_PROOFS:
+        return False
+
+    # Verify all execution proofs
+    for proof in execution_proofs:
+        # Get EL program from proof system ID
+        # TODO: The proof system ID is really an ID for a proof system and EL combination
```

Comment:
TODO is noting that proofID has a 1-1 mapping with subnet which is really a 1-1 mapping between EL and zkEVM combinations. Not just the proof system being used